### PR TITLE
Quexington's Offer generalization bringup

### PR DESCRIPTION
**Pull request by AmineKhaldi**
This is part of PR #11557 to be reviewed separately for convenience.

### CODE REVIEW

**Comment by AmineKhaldi** [comment]
@Quexington @mariano54 this is an effort to bring-in parts of nft work that can be reviewed/merged separately. Your reviews are appreciated.

**Review by Quexington — COMMENTED** [review]
I wrote this code so I'm not sure I should be the one to give it the thumbs up.  This is basically a refactor of the offer code so if the existing tests are working then that gives me pretty high confidence.

File: chia/rpc/wallet_rpc_client.py
Comment (Quexington) [inline chia/rpc/wallet_rpc_client.py:424]: To fix the mypy error this should be Dict[Union[uint32, str], int]

**Comment by AmineKhaldi** [comment]
Closing to reopen, the checks seem stuck.

**Review by Quexington — COMMENTED** [review]
:+1: to Amine's merge skills, but someone else should approve this before it gets merged into main

**Review by xdustinface — CHANGES_REQUESTED** [review]
@Quexington Left some comments, some of them should be addressed here, others maybe not (yet?). Overall it would be nice to see more tests for all the new introduced code and logic!

The merge seems fine, i got the same result when picking the PRs into `main` after resolving the conflicts.

File: chia/wallet/cat_wallet/cat_outer_puzzle.py
Comment (Quexington) [reply chia/wallet/cat_wallet/cat_outer_puzzle.py:21]: This is covered by the tests that already exist for offers.  This code was scattered throughout the offer driver before and has just been extracted to this file now.

File: chia/wallet/trade_manager.py
Comment (Quexington) [reply chia/wallet/trade_manager.py:51]: This is a good idea, I knew something like that must be possible but I didn't know how to do it.

File: chia/wallet/trading/offer.py
Comment (xdustinface) [reply chia/wallet/trading/offer.py]: It's just a (at least imo) simpler statement and you update the dict instead of creating a new one each time. Like you would use `dict["new_key"] = new_value` instead of `dict = {**dict, "new_key": new_value}`.

File: chia/wallet/cat_wallet/cat_outer_puzzle.py
Comment (mariano54) [inline chia/wallet/cat_wallet/cat_outer_puzzle.py:42]: this is not Optional right?

File: chia/wallet/cat_wallet/cat_wallet.py
Comment (mariano54) [inline chia/wallet/cat_wallet/cat_wallet.py:827]: This will also raise if there are enough confirmed balance, but not enough spendable balance. Is that OK?

File: chia/wallet/wallet.py
Comment (mariano54) [inline chia/wallet/wallet.py:536]: same here, this can raise if there is not enough spendable funds

**Review by mariano54 — COMMENTED** [review]
These changes seem good, but I am not deeply familiar with offer code so I can't provide much useful feeback.
Looks like no slow tests are added. We should ensure all tests pass and without flakiness.

File: chia/wallet/trade_manager.py
Comment (xdustinface) [inline chia/wallet/trade_manager.py:293]:
```suggestion
```

Its only passed into `_ create_offer_for_ids` here where it gets created if its `None`.

**Review by xdustinface — CHANGES_REQUESTED** [review]
@Quexington After the "avoid mutation of the default `driver_dict` value" changes the test `test_trades.py` is failing here https://github.com/Chia-Network/chia-blockchain/blob/71a8ac2e698726d4b7d43832be382c161c3ecf8c/tests/wallet/cat_wallet/test_trades.py#L199

I debuged into it a bit and it fails here 

https://github.com/Chia-Network/chia-blockchain/blob/22919901f1a1556194c74773488e1dd15e4abe3b/chia/wallet/trading/offer.py#L124-L131

because the `asset_id` calculated with the `puzzle_driver` is not in `driver_dict`. I assume its related to the fact that we before had the drivers of the previous offer creations in `driver_dict` always because it always used the mutated default value. And the code doesn't seem to generate the correct one by itself now. Can you look into it?

File: chia/wallet/cat_wallet/cat_outer_puzzle.py
Comment (Quexington) [reply chia/wallet/cat_wallet/cat_outer_puzzle.py:42]: For this class it's not, but in general it's possible to have one of these classes that does return None.  These should probably be implemented as `Protocol`s but I didn't know about those until Kevin's comments on this PR.

File: chia/wallet/cat_wallet/cat_wallet.py
Comment (Quexington) [reply chia/wallet/cat_wallet/cat_wallet.py:827]: Yeah we don't want to offer coins that aren't spendable

File: chia/wallet/wallet.py
Comment (Quexington) [reply chia/wallet/wallet.py:536]: ditto to above

File: tests/clvm/test_puzzle_drivers.py
Comment (xdustinface) [inline tests/clvm/test_puzzle_drivers.py:27]:
```suggestion
assert puzzle_info.type() == "TEST"
```

**Review by xdustinface — CHANGES_REQUESTED** [review]
@Quexington So you fixed the tests by just providing the infos for all CATs in the `driver_dict`. Correct me if im wrong but from my understanding it should work even with no `driver_dict` provided given the fact that the wallet has `get_asset_id` and `get_puzzle_info` implemented?

Also im not sure but it looks to me like we have not all of the cases in `create_offer_for_ids` tested. Like for offering/requesting both with wallet_id/asset_id, with and without `driver_dict` and also test with an invalid drivers which differs from the the `get_puzzle_info` result. 

Ah and btw, `pre-commit` is still unhappy about workflows and some `isort` fun :)

**Comment by xdustinface** [comment]
@Quexington The issue is the indentation of https://github.com/Chia-Network/chia-blockchain/blob/79ddd63ac830fe49b79726d6dfa898953156a37e/chia/wallet/trade_manager.py#L372-L382 It needs to be shifted one to the left so that it adds the driver info for each entry of the offer. Right now it only adds it for the last one. So you can revert your "just provide all drivers" changes to the test but like i said it would be nice to see all those cases tested in here :)

**Comment by xdustinface** [comment]
@Quexington Can you explain why we even need to allow providing the `driver_dic` in `_create_offer_for_ids` and also in the RPC interface since from my understanding the driver infos can be fetched in both cases like:

- Offer creation: By calling the wallet functions
- Offer taking: By puzzle matching from the spend bundle

Do i miss something here?

**Review by xdustinface — CHANGES_REQUESTED** [review]
@Quexington See the comments about `__eq__`.
@AmineKhaldi There is a conflict, see https://github.com/AmineKhaldi/chia-blockchain/pull/1

File: chia/wallet/puzzle_drivers.py
Comment (xdustinface) [reply chia/wallet/puzzle_drivers.py:43]: Oh i see, can you then you also add tests for this to test the key and value error cases? I got no test failure when removing your `__eq__` implementations.

**Review by xdustinface — APPROVED** [review]

**Comment by emlowe** [comment]
I have no problems with this, just double-checked with Amine that this is what he wants to do - and Amine is good. So I'm therefore also good.

**Review by wjblanke — APPROVED** [review]
aok

**Pull request merged**
